File: core/controllers/secure.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import OAuth2PasswordRequestForm
from typing import List, Tuple

from services.search import search
from services.story import story
from services.user import user, personalization, auth, feedback, favorite
from services import redis
from shared.types import story_types, search_types

router = APIRouter()


@router.get("/users/me")
async def read_users_me(current_user: user.User = Depends(
    auth.get_current_active_user)):
  return {"user": current_user.dict()}


@router.put("/users")
def updating_user(current_user: user.User,
                  user_id: str = Depends(auth.get_current_user_id)):
  current_user.user_id = user_id
  updated_user = user.update(current_user)
  if type(updated_user) == str:
    raise HTTPException(status_code=400, detail=updated_user)
  return {"user": updated_user.dict()}


@router.get("/users/email")
async def confirm_email(current_user: user.User = Depends(
    auth.get_current_active_user)):
  current_user.is_email_subscribed = True
  user.update(current_user)
  return {"user": current_user.dict()}


@router.get("/google/{token}")
async def google_auth(token: str):
  current_user = auth.google_token_check(token)
  if type(current_user) == str:
    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail=current_user,
        headers={"WWW-Authenticate": "Bearer"},
    )
  access_token = auth.create_access_token(data={"sub": current_user.user_id})
  return {"token": access_token, "user": current_user.dict()}


@router.post("/token")
async def login_for_access_token(
    form_data: OAuth2PasswordRequestForm = Depends()):
  current_user = auth.authenticate_user(form_data.username, form_data.password)
  if not current_user:
    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Incorrect username or password",
        headers={"WWW-Authenticate": "Bearer"},
    )
  access_token = auth.create_access_token(data={"sub": current_user.user_id})
  return {"token": access_token, "user": current_user.dict()}


@router.get("/user/info/{language}",
            response_model=personalization.Personalization)
async def get_personalization_items(language: str,
                              user_id: str = Depends(
                                  auth.get_current_user_id)):
  return personalization.get_personalization(user_id, language)


@router.get("/recommended-news/{language}",
            response_model=Tuple[story_types.ShortStory, ...])
def get_recommended_stories(language: str,
                            user_id: str = Depends(auth.get_current_user_id)):
  return story.get_recommended_stories(user_id, language)


@router.post("/single-article/{language}",
             response_model=story_types.ShortStory)
def get_single_story(language: str,
                     not_id_list: List[str],
                     user_id: str = Depends(auth.get_current_user_id)):
  return story.get_single_story(not_id_list, language)


@router.post("/feedback")
def receive_feedback(user_feedback: feedback.UserFeedback,
                     user_id: str = Depends(auth.get_current_user_id)):
  user_feedback.user_id = user_id
  redis.worker_queue.enqueue(feedback.received, user_feedback)
  return {"job": "received Queued"}


@router.post("/source")
def update_from_user_source(fave: favorite.Favorite,
                            user_id: str = Depends(auth.get_current_user_id)):
  fave.user_id = user_id
  result = favorite.update_from_user(fave, "FavoriteSource")
  return {"result": result}


@router.post("/author")
def update_from_user_author(fave: favorite.Favorite,
                            user_id: str = Depends(auth.get_current_user_id)):
  fave.user_id = user_id
  result = favorite.update_from_user(fave, "FavoriteAuthor")
  return {"result": result}


@router.post("/keyword")
def update_from_user_keyword(fave: favorite.Favorite,
                             user_id: str = Depends(auth.get_current_user_id)):
  fave.user_id = user_id
  result = favorite.update_from_user(fave, "FavoriteKeyword")
  return {"result": result}


@router.post("/entity")
def update_from_user_entity(fave: favorite.Favorite,
                            user_id: str = Depends(auth.get_current_user_id)):
  fave.user_id = user_id
  result = favorite.update_from_user(fave, "FavoriteEntity")
  return {"result": result}


@router.post("/search", response_model=List[story_types.ShortStory])
def get_stories(search_query: search_types.SearchQuery,
                user_id: str = Depends(auth.get_current_user_id)):
  search_query.user_id = user_id
  return search.simple_search(search_query)


@router.post("/search/personalized",
             response_model=List[story_types.ShortStory])
def get_personalized_stories(search_query: search_types.SearchQuery,
                             user_id: str = Depends(auth.get_current_user_id)):
  search_query.user_id = user_id
  return search.solr_search_with_personalization(search_query)

File: core/internal/controllers/secure.go
package controllers

import (
	"log"
	"net/http"

	"core/internal/feedback"
	"core/internal/story"
	"core/internal/tasks"
	"core/internal/user"

	"github.com/golang-jwt/jwt"
	"github.com/google/uuid"
	"github.com/gorilla/mux"
	"github.com/hibiken/asynq"
)

func SecureRoutes(r *mux.Router, queueClient *asynq.Client) *mux.Router {

	r.HandleFunc("/users/me", Chain(func(w http.ResponseWriter, r *http.Request) {
		userId := TokenSubFromRequest(r)
		currentUser, err := user.GetUserById(userId)
		if err != nil {
			RespondWithError(w, 500, "User not found")
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]user.User{"user": currentUser})
	}, Secure(), Method("GET"))).Methods("GET", "OPTIONS")

	r.HandleFunc("/users", Chain(func(w http.ResponseWriter, r *http.Request) {
		userId := TokenSubFromRequest(r)
		var updateUser user.UserUpdate
		ok := DecodeJSON(w, r, &updateUser)
		if !ok {
			return
		}
		if updateUser.UserId != userId {
			RespondWithError(w, 401, "You may only update your user")
			return
		}
		currentUser, err := user.UpdateUser(updateUser)
		if err != nil {
			RespondWithError(w, 500, "Failed to updated user")
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]user.User{"user": currentUser})
	}, Secure(), Method("PATCH"))).Methods("PATCH", "OPTIONS")

	r.HandleFunc("/users/email", Chain(func(w http.ResponseWriter, r *http.Request) {
		userId := TokenSubFromRequest(r)
		update := true
		updateUser := user.UserUpdate{
			UserId:            userId,
			IsEmailSubscribed: &update,
		}
		updatedUser, err := user.UpdateUser(updateUser)
		if err != nil {
			RespondWithError(w, 500, "Failed to updated user")
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]user.User{"user": updatedUser})
	}, Secure(), Method("GET"))).Methods("GET", "OPTIONS")

	r.HandleFunc("/google/{token}", Chain(func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		token := vars["token"]
		currentUser, err := user.GoogleTokenCheck(token)
		if err != nil {
			RespondWithError(w, 401, "Failed token check")
			return
		}
		accessToken, er := user.CreateAccessToken(jwt.MapClaims{"sub": currentUser.UserId})
		if er != nil {
			RespondWithError(w, 500, er.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]any{"token": accessToken, "user": currentUser})
	}, Method("GET"))).Methods("GET", "OPTIONS")

	r.HandleFunc("/token", Chain(func(w http.ResponseWriter, r *http.Request) {
		username := r.FormValue("username")
		password := r.FormValue("password")
		currentUser, err := user.AuthenticateUser(username, password)
		if err != nil {
			RespondWithError(w, 401, "Failed to Authenticate")
			return
		}
		accessToken, er := user.CreateAccessToken(jwt.MapClaims{"sub": currentUser.UserId})
		if er != nil {
			RespondWithError(w, 500, "Failed to create access token")
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]any{"token": accessToken, "user": currentUser})
	}, Method("POST"))).Methods("POST", "OPTIONS")

	r.HandleFunc("/user/info/{language}", Chain(func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		userId := TokenSubFromRequest(r)
		personalization, err := user.GetPersonalization(userId, language)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, personalization)
	}, Secure(), Method("GET"))).Methods("GET", "OPTIONS")

	r.HandleFunc("/recommended-news/{language}", Chain(func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		userId := TokenSubFromRequest(r)
		stories, err := story.GetRecommendedStories(userId, language)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, stories)
	}, Secure(), Method("GET"))).Methods("GET", "OPTIONS")

	r.HandleFunc("/single-article/{sourceId}/{language}", Chain(func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		sourceId := vars["sourceId"]
		var notIdList []uuid.UUID
		ok := DecodeJSON(w, r, &notIdList)
		if !ok {
			return
		}
		story, err := story.GetSingleStory(notIdList, sourceId, language)
		if err != nil {
			RespondWithError(w, 204, "No story found")
			return
		}
		RespondWithJSON(w, http.StatusOK, story)
	}, Secure(), Method("POST"))).Methods("POST", "OPTIONS")

	r.HandleFunc("/feedback", Chain(func(w http.ResponseWriter, r *http.Request) {
		userId := TokenSubFromRequest(r)
		var userFeedback feedback.UserFeedback
		ok := DecodeJSON(w, r, &userFeedback)
		if !ok {
			return
		}
		userFeedback.UserId = userId
		task, err := tasks.NewStoreUserFeedbackTask(userFeedback)
		if err != nil {
			RespondWithError(w, 500, "Could not create NewStoreUserFeedbackTask")
			return
		}
		info, err := queueClient.Enqueue(task)
		if err != nil {
			RespondWithError(w, 500, "Could not enqueue NewStoreUserFeedbackTask")
			return
		}
		log.Printf("enqueued task: id=%s queue=%s", info.ID, info.Queue)
		RespondWithJSON(w, http.StatusOK, map[string]string{"StoreUserFeedback": "Job Queued"})
	}, Secure(), Method("POST"))).Methods("POST", "OPTIONS")

	r.HandleFunc("/favorite/{collection}", Chain(func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		collection := vars["collection"]
		dbCollection, err := user.GetFavoriteCollection(collection)
		if err != nil {
			RespondWithError(w, 404, "Favorite"+collection+" not found")
			return
		}
		var favorite user.Favorite
		ok := DecodeJSON(w, r, &favorite)
		if !ok {
			return
		}
		userId := TokenSubFromRequest(r)
		favorite.UserId = userId
		er := user.UpdateFavoriteFromUser(favorite, dbCollection)
		if er != nil {
			RespondWithError(w, 500, "Failed to update favorite "+collection)
			return
		}
		RespondWithJSON(w, http.StatusOK, map[string]string{"result": "Updated user favortie " + collection})
	}, Secure(), Method("POST"))).Methods("POST", "OPTIONS")

	return r
}
